lib/ansible/modules/cloud/vmware/vmware_guest_controller.py

```diff
@@ -0,0 +1,551 @@
+     - This is a required parameter, if parameter C(name) or C(moid) is not supplied.
+     type: str
+   moid:
+     description:
+     - Managed Object ID of the instance to manage if known, this is a unique identifier only within a single vCenter instance.
+     - This is required if C(name) or C(uuid) is not supplied.
+     type: str
+   folder:
+     description:
+     - Destination folder, absolute or relative path to find an existing guest.
+     - This is a required parameter, only if multiple VMs are found with same name.
+     - The folder should include the datacenter. ESX's datacenter is ha-datacenter
+     - 'Examples:'
+     - '   folder: /ha-datacenter/vm'
+     - '   folder: ha-datacenter/vm'
+     - '   folder: /datacenter1/vm'
+     - '   folder: datacenter1/vm'
+     - '   folder: /datacenter1/vm/folder1'
+     - '   folder: datacenter1/vm/folder1'
+     - '   folder: /folder1/datacenter1/vm'
+     - '   folder: folder1/datacenter1/vm'
+     - '   folder: /folder1/datacenter1/vm/folder2'
+     type: str
+   datacenter:
+     description:
+     - The datacenter name to which virtual machine belongs to.
+     default: ha-datacenter
+     type: str
+   use_instance_uuid:
+     description:
+     - Whether to use the VMware instance UUID rather than the BIOS UUID.
+     default: no
+     type: bool
+   controllers:
+     description:
+     - A list of disk or USB controllers to add or remove.
+     - Total 4 disk controllers with the same type are allowed per VM.
+     - Total 2 USB controllers are allowed per VM, 1 USB 2.0 and 1 USB 3.0 or 3.1.
+     - For specific guest OS, supported controller types please refer to VMware Compatibility Guide.
+     suboptions:
+       controller_number:
+         description:
+         - Disk controller bus number. When C(state) is set to C(absent), this parameter is required.
+         - When C(type) set to C(usb2) or C(usb3), this parameter is not required.
+         type: int
+         choices:
+           - 0
+           - 1
+           - 2
+           - 3
+       type:
+         description:
+         - Type of disk or USB controller.
+         - From vSphere 6.5 and virtual machine with hardware version 13, C(nvme) controller starts to be supported.
+         required: true
+         type: str
+         choices:
+           - buslogic
+           - lsilogic
+           - lsilogicsas
+           - paravirtual
+           - sata
+           - nvme
+           - usb2
+           - usb3
+       state:
+         description:
+         - Add new or remove existing controllers operations.
+         - If C(state) is set to C(absent), controller will be removed from virtual machine when there is no disk or device attaching to it.
+         - If C(state) is set to C(new), controller will be added if the number of given controller type not reach the maximum.
+         required: true
+         type: str
+         choices:
+           - new
```

Comment:
This will not be idempotent, is that intended?  We should try to support `present`, or if this is not possible warn users that there is no idempotency.

Reply:
C(new) will add controllers specified until reach it's maximum if executed several times, then will always return 'can not add any more'.
C(absent) will remove the specified controller at the first time, then if execute more times it will always return 'not found controller'.
If add C(present), it will do nothing.

Reply:
Ok then we should have a message in the docs then to warn users that this module/option will not be idempotent, so that it doesn't catch them unawares.

Reply:
Ok, thanks, will add this in the doc.

Reply:
I don't like the `new` state too. This will be confusing for the users since it's a rather non-standard behaviour. And worst, this breaks the idem-potency of the module.

May I suggest something like:

```yaml
- name: Set the disk of the VM
  vmware_guest_controller:
    name: test_VM
    controllers:
      - state: present
        type: sata
      - state: present
        type: sata
      - state: present
        type: nvme
      - state: present
        type: usb3
```

Reply:
Ok, I changed `new` to `present`. Thanks for the reviewing.